lib/util/location.ex

```diff
@@ -115,76 +223,126 @@ defmodule Util.Location do
-          # then the nearest point on the segment (0, 0) -- (1, 0) is the
-          # point (u, 0). Otherwise, it's whichever is closer of (0, 0) or
-          # (1, 0).
-          nearest_u =
-            cond do
-              u < 0 -> 0
-              u > 1 -> 1
-              true -> u
-            end
-
-          # Transform the nearest point back to (x, y) space
-          {nearest_u * segment_end_x, nearest_u * segment_end_y}
+    # Convert both the segment end-point and the off-segment point
+    # into vectors representing their displacements from the segment
+    # start-point. For example,
+    #
+    #                      P <-- off-segment point
+    #                     /
+    #                    /
+    #                   /
+    #  start-point --> S--------E <-- end-point
+    #
+    # segment_vector would be the vector pointing from S to E
+    # point_vector would be the vector pointing from S to P
+    #
+    # Ww don't compute it because we don't need to, but (0.0, 0.0)
+    # would be a vector pointing from S to itself.
+    segment_vector = displacement_from(end_loc, start_loc)
+    point_vector = displacement_from(point, start_loc)
+
+    # For ease of computation, we first find the nearest vector - that
+    # is, a vector pointing from the start-point to the closest point
+    # that we're looking for. Then, as the last step, we'll use that
+    # vector to displace from the start-point to find the lat/long
+    # coordinates of the final result
+    nearest_vector =
+      case Vector2d.dot_product(segment_vector, segment_vector) do
+        # If the segment happens to be zero-length (that is, if the
+        # start and end points coincide), then the nearest point on
+        # the segment is the segment start-point (and the end-point,
+        # since they're the same point), so the vector displacement
+        # from the start-point will be zero. For example, in the
+        # diagram below, regardless of where P is, S (and E) is the
+        # closest point, because it's the only point.
+        #
+        #      P
+        #     /
+        #    /
+        #   /
+        #  SE
+        #
+        0.0 ->
+          Vector2d.zero()
+
+        # Otherwise we project the off-segment vector onto the line
+        # containing the segment, and figure the nearest point from
+        # there. For example, in the diagrams below, Q is what you get
+        # by projecting P onto the line. Notice that Q doesn't always
+        # fall onto the actual segment, because the segment doesn't go
+        # on forever.
+        #
+        #           P                           P
+        #          /                             \
+        #         /                               \
+        #        /                                 \
+        #       S---Q----E                      Q   S--------E
+        #
+        squared_segment_length ->
+          projected_vector = Vector2d.project_onto(point_vector, segment_vector)
```

Comment:
You're right - I should have said something like `S -> Q is what you get when you project S -> P onto the line containing S -> E`. 

Literally _everything_ except for the lat/long coordinates is vectors, not points (you can kind of tell that in the code, because almost all of the variables inside the function are `%Vector2d{...}`'s, but that's only immediately obvious if you look at the `@spec`'s of the various functions called), but I was fuzzy in my language in the explanation. I shall clean that up before merging.